src/preamble/01_what_do_we_mean_when_we_say_cabal.md

```diff
@@ -1 +1,73 @@
 # What do we mean when we say cabal?
+
+Cabal is an umbrella term that can refer to either cabal-the-spec (.cabal
+files), cabal-the-library (code that understands .cabal files), or
+cabal-the-tool (the cabal-install package which provides the cabal executable).
+
+## tl;dr
+
+1. [cabal-install](#cabalInstall) is the command line utility to help configure,
+   compile and install Haskell libraries and programs.
+
+2. [The Cabal build system](#cabailBuildSystem), a specification for defining
+   Haskell packages, together with a library for performing builds. This build
+   system is used by cabal-install and other package managers (stack for eg).
+   For eg: parsing .cabal files (among others).
+
+3. [.cabal file](#cabalFile) This is the format that specifies the content of
+   Haskell packages.
+
+## The .cabal file<a name="cabalFile"></a>
+
+The .cabal file contains information that drives the compilation and building of
+Haskell packages. This is the format that specifies the content of Haskell
+packages. It is a text-based, key-value format, that is divided into subsections
+called stanzas. Each section consists of a number of property descriptions in
+the form of field/value pairs, with a syntax roughly like mail message headers.
+It is denoted with a file with the .cabal extension.
+
+This file should contain a number global property descriptions and several
+sections. The package properties describe the package as a whole, such as name,
+license, author, etc. Optionally, a number of configuration flags can be
+declared. These can be used to enable or disable certain features of a package.
+The (optional) library section specifies the library properties and relevant
+build information. Following is an arbitrary number of executable sections which
+describe an executable program and relevant build information.
```

Comment:
I think this paragraph needs to be tweaked. We need to be very careful about being exceedingly clear here. This is one of the more complicated aspects of cabal.

My suggestions are:
- Start with a sentence explaining what package properties are (perhaps leave out the concept of being "global" they are really just top level fields). Make sure to explain that package properties are metadata about the package that will be built as a result of calling `cabal build`, usually the `.cabal` file lives at the root directory that contains the haskell source code corresponding to the package.
- Explain that the optional component stanzas (library, executable, test-suite, and benchmark) are things called components. Perhaps link to another section of the user guide that covers them in more detail.
- If you are going to mention more specific fields like configuration flags, then make sure to be very clear about what they can do (i.e. enable language extensions, or apply flags to ghc)

Reply:
I'll rewrite this paragraph.

Reply:
@JonathanLorimer 
I've rewritten this para as: 

The .cabal file contains information that drives the compilation and building of
Haskell packages. Usually the .cabal file lives at the root directory that
contains the haskell source code corresponding to the package. Usually this file
is named as `project-name.cabal`. It is a text-based, key-value format, that is
divided into subsections called stanzas. Each section consists of a number of
property descriptions.

The package properties describe the package as a whole, such as name, license,
author, dependenices, etc. It also contains optional information about optional
components such as
[library properties](../new_to_cabal/06_first_cabal_library.md),
[executables](../new_to_cabal/07_first_cabal_executable.md),
[test-suite](../leveling_up/02_first_cabal_test-suite.md),
[benchmark](src/leveling_up/03_first_cabal_benchmark.md). These components are
also called stanzas.

One of the purposes of Cabal is to make it easier to build a package with
different Haskell implementations. So it provides abstractions of features
present in different Haskell implementations and wherever possible it is best to
take advantage of these to increase portability. For example one of the pieces
of information an author can put in the package’s .cabal file is what language
extensions the code uses. This is far preferable to specifying flags for a
specific compiler as it allows Cabal to pick the right flags for the Haskell
implementation that the user picks. It also allows Cabal to figure out if the
language extension is even supported by the Haskell implementation that the user
picks. Where compiler-specific options are needed however, there is an “escape
hatch” available. The developer can specify implementation-specific options and
more generally there is a configuration mechanism to customise many aspects of
how a package is built depending on the Haskell implementation, the Operating
system, computer architecture and user-specified configuration flags.


Is the last para fine, or should I reduce it?